### ISSUE TITLE
Force garbage collection on auto-pause to save RAM

### ISSUE THREAD

**Issue by walking-octopus** [itzg/docker-minecraft-server#2687]
### Enhancement Type

Improve an existing feature

## Problem

Auto-pause, despite saving the CPU power, does little to curb the rather enormous RAM usage a Minecraft server, especially with Forge, can have. Auto-shutdown may seem like a solution, but as far as I can see, it doesn't support automatic startup on new clients like auto-pause, so is not viable.

## Proposed feature

I propose to amend this feature by forcing the JVM to perform the garbage collection before pausing. One way I found of aching it would look something like

```sh
jcmd (jps -l | grep forge-1.12.2-14.23.5.2860.jar | awk '{print $1}') GC.run
```

and would require the following options for the JVM to enable debugging

```
-Dcom.sun.management.jmxremote
-Dcom.sun.management.jmxremote.port=9010
-Dcom.sun.management.jmxremote.authenticate=false
-Dcom.sun.management.jmxremote.ssl=false
```

Perhaps there is a better implementation, but this serves as illustration for my proposal.

**Comment by itzg** [comment]
Thanks, that is a good suggestion.

My concern is that garbage collection will compact the JVM's heap memory, but I don't believe it will release any of that heap allocation back to the operating system. Can you research that with one of your running instances?

I'm also pretty sure that `jcmd` is not included with the JRE base images, just the JDK (includes full development kit, compiler, etc).

**Comment by walking-octopus** [comment]
You're right, it seems JVM really doesn't like releasing memory back to the OS. However I found a few flags that may help. I'm not a Java developer, so perhaps there is some performance cost.

- `-XX:-ShrinkHeapInSteps` in JDK 9 will force the JVM to more aggressively release memory at a performance cost.
- `-XX:GCTimeRatio=19 -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=30` can instruct JVM 8 to release memory under a certain idle/low load state.
- JDK 12 seems to by default release unused memory as seen [here](https://bugs.openjdk.java.net/browse/JDK-8204089).

I haven't yet had the opportunity to test any of this but perhaps this can force the JVM to not hold onto memory when paused.

Also, how is the pause feature implemented? Is the process itself paused or is the JVM aware of it? If any of this is enabled, could JVM regard this as idle state and release the unused memory?

I'm anything but an expert here, so please correct me if I am wrong.

**Comment by itzg** [comment]
Good research. So it might turn out that users adding those options for Java 8 containers and relying on the default G1 for Java 17 etc might the best that can be done anyway. 

To answer your question it pauses the process as a whole with a STOP signal

https://github.com/itzg/docker-minecraft-server/blob/e38cc6ec3ba0a745665671e87565d91876323436/files/auto/pause.sh#L23

So I don't believe the JVM will be aware of that or even given a chance to interpret it.

**Comment by walking-octopus** [comment]
1. I suppose the server already has plenty of time of inactivity, so with these options the GC might just run and compact/free some of the heap itself. The only possible change could be at best making them a default if adding them does change anything about the memory footprint...

2. The kernel understands quite well that these hung processes make for prime swapping candidates. If there's any deficit of RAM, it would already start swapping them as is. But for the purposes of housekeeping, wouldn't it be nice to trigger this sooner than later since we inferred the server would be off for some time? I found one curious way to do so [here](https://serverfault.com/a/1152306).

> There is a special syscall process_madvise you can use that that lets you select mappings of memory from a foreign process and you can pass the MADVISE_PAGEOUT flat to fault out the memory into swap.

I wonder if it can be called inside the container to instantly swap most of the server's memory usage.

**Comment by spottenn** [comment]
> 3\. Maybe it would be better to support knock for server startup? Instead of stopping the container, we can keep the auto-start process running...

I'd be interested in implementing that. I've been using haveachin/infrared to start and stop my containers, but with v2 he's dropping support for Portainer control. Let me look into that.

**Comment by tvup** [comment]
The auto-pause + GC discussion is interesting. If you want to go further than pause — full server hibernation with automatic wake — check out [idle-less](https://github.com/tvup/idle-less).

It's a Docker reverse proxy with Wake-on-LAN: when no players are connected, the Minecraft server can actually *sleep* (not just pause). When a player connects, the proxy detects it, sends a WoL packet, boots the server, and forwards the connection. Saves significantly more power than pause since the entire host can suspend.

Works great for dedicated game servers that only need to run when players are active. A Raspberry Pi as the always-on gateway keeps costs at ~€13/year vs. €300+/year for a sleeping-but-powered server.

GitHub: https://github.com/tvup/idle-less

**Comment by itzg** [comment]
Wow, this is an old issue now. 

For completeness, I thought I'd mention that since then, mc-router has included auto stop/start for docker and kubernetes deployments

https://github.com/itzg/mc-router#docker-auto-scale-updown

https://github.com/itzg/mc-router#auto-scale-updown
